pytest_operator/plugin.py

```diff
@@ -500,37 +732,83 @@ async def create_crash_dump(self) -> bool:
+        If the model is not marked as kept, ops_test will destroy the model.
 
-        # NOTE (rgildein): Create juju-crashdump only if any tests failed,
-        # `juju-crashdump` flag is enabled and OpsTest.keep_model == False
-        if (
-            self.request.session.testsfailed > 0
-            and self.crash_dump
-            and self.keep_model is False
-        ):
-            await self.create_crash_dump()
+        @param                   str alias: alias of the model
+        @param Optional[float,int] timeout: how long to wait for it to be removed
+        """
+        if not self._controller:
+            log.error("No access to controller, skipping...")
+            return
 
-        if not self.keep_model:
-            # Forcibly destroy machines in case any units are in error.
-            for machine in self.model.machines.values():
-                try:
-                    log.info(f"Destroying machine {machine.id}")
-                    await machine.destroy(force=True)
-                except DeadEntityException as e:
-                    log.warning(e)
-                    log.warning("Machine already dead, skipping")
-            await self.model.disconnect()
-            log.info(f"Destroying model {self.model_name}")
-            await self._controller.destroy_model(self.model_name)
-        else:
-            await self.model.disconnect()
-        if self._controller:
-            await self._controller.disconnect()
+        if alias not in self.models:
+            log.warning(f"No access to model alias {alias}, skipping...")
+            return
+
+        with self.model_context(alias) as model:
+            await self.log_model()
+
+            # NOTE (rgildein): Create juju-crashdump only if any tests failed,
+            # `juju-crashdump` flag is enabled and OpsTest.keep_model == False
+            if (
+                self.request.session.testsfailed > 0
+                and self.crash_dump
+                and self.keep_model is False
+            ):
+                await self.create_crash_dump()
+
+            if not self.keep_model:
+                # Forcibly destroy machines in case any units are in error.
+                for machine in model.machines.values():
+                    try:
+                        log.info(f"Destroying machine {machine.id}")
+                        await machine.destroy(force=True)
+                    except DeadEntityException as e:
+                        log.warning(e)
+                        log.warning("Machine already dead, skipping")
+                await model.disconnect()
+                log.info(f"Destroying model {self.model_name}")
+                await self._controller.destroy_model(self.model_name)
+                log.info(f"Waiting on model teardown {self.model_name}...")
+                await asyncio.wait_for(
+                    self._model_gone(self.model_name), timeout=timeout
+                )
```

Comment:
This isn't wrong, but there's a mix of using `self.model_name` and `model` (not `self.model`) here that made me wonder if I understood how `self.model_context()` worked.  Just something to consider